lib/hls/hls_parser.js

```diff
@@ -985,14 +990,56 @@ shaka.hls.HlsParser = class {
       goog.asserts.assert(streamInfos.audio.length ||
           streamInfos.video.length, 'We should have created a stream!');
 
+      const drmInfos = [];
+      const keyIds = new Set();
+
+      if (sessionKeyTags.length > 0) {
+        const characteristics = tag.getAttributeValue('CHARACTERISTICS');
+        for (const drmTag of sessionKeyTags) {
+          const drmCharacteristics =
+              drmTag.getAttributeValue('CHARACTERISTICS');
+          if ((characteristics && !drmCharacteristics) ||
+              (!characteristics && drmCharacteristics) ||
+              (drmCharacteristics && characteristics &&
+              !characteristics.includes(drmCharacteristics))) {
+            continue;
+          }
+          const method = drmTag.getRequiredAttrValue('METHOD');
+          if (method != 'NONE' && method != 'AES-128') {
+            // According to the HLS spec, KEYFORMAT is optional and implicitly
+            // defaults to "identity".
+            // https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-11#section-4.4.4.4
+            const keyFormat =
+                drmTag.getAttributeValue('KEYFORMAT') || 'identity';
+            const drmParser =
+                shaka.hls.HlsParser.KEYFORMATS_TO_DRM_PARSERS_[keyFormat];
+
+            const drmInfo = drmParser ?
+                drmParser(drmTag, /* mimeType= */ '') : null;
```

Comment:
Instead of parsing the same session key tag over and over for each variant, could you parse them once and push the drm infos down to the variants?

Reply:
Changed!